service/src/main/java/com/myhome/services/springdatajpa/UserSDJpaService.java

```diff
@@ -76,6 +88,48 @@
     }).orElse(Optional.empty());
   }
 
+  @Override
+  public void requestResetPassword(ForgotPasswordRequest forgotPasswordRequest) {
+    if (forgotPasswordRequest.getEmail() != null) {
+      SecurityToken newSecurityToken = securityTokenService.createPasswordResetToken();
+      Optional<User> userOptional = userRepository.findByEmailWithTokens(forgotPasswordRequest.getEmail());
+      userOptional.map(user -> {
+        user.getUserTokens().add(newSecurityToken);
+        userRepository.save(user);
+        mailService.sendPasswordRecoverCode(user, newSecurityToken.getToken());
+        return user;
+      });
+    }
+  }
+
+  @Override
+  public boolean resetPassword(ForgotPasswordRequest passwordResetRequest) {
+    if (passwordResetRequest != null) {
+      Optional<User> userOptional = userRepository.findByEmailWithTokens(passwordResetRequest.getEmail());
+      return userOptional.map(user -> {
+        SecurityToken userPasswordResetToken = user.getUserTokens()
+            .stream()
+            .filter(token -> !token.isUsed()
+                && token.getTokenType() == SecurityTokenType.RESET
+                && token.getToken().equals(passwordResetRequest.getToken())
+                && token.getExpiryDate().isAfter(LocalDate.now()))
+            .findFirst()
+            .orElse(null);
+        if (userPasswordResetToken != null) {
```

Comment:
Please think about a more functional way to do that (you have `.map` operator, optional.

Reply:
I cannot use the `user` variable inside `userPasswordResetToken.map(...)` because it is not `final`

Reply:
I cannot use the 'user' variable inside `userPasswordResetToken.map` because it is not final

Reply:
if I wrap `passwordResetRequest` in optinal then I can't use `passwordResetRequest.getToken()` inside `passwordResetRequest.map(...)` because it isn't final